test/test_pid_reuse.c

```diff
@@ -0,0 +1,167 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/syscall.h>
+#include <assert.h>
+#include <errno.h>
+#include <pthread.h>
+#include "zlib.h"
+
+/* Globals shared by all threads. */
+z_stream c_stream;
+
+void* comp_file_multith (void *iterations)
+{
+	int err;
+	long *sleep_sec;
+
+	pid_t my_pid = syscall(SYS_gettid);
+	fprintf(stderr, "thread started: pid: %d\n", (int) my_pid);
+
+	/* Open the window. */
+	err = deflateInit(&c_stream, Z_DEFAULT_COMPRESSION);
+	if (err != 0) {
+		printf("nx_deflateInit err %d\n", err);
+		return (void *) 1;
+	}
+
+	sleep_sec = (long *) iterations;
+	sleep(sleep_sec[0]);
+
+	fprintf(stderr, "thread %d exits\n", (int) my_pid);
+	return (void *) 0;
+}
+
+void fork_children (int parent_pid) {
+	int child_id, i = 0;
+	int seconds = 10;
+	int children = 2048;
+
+	for (i=0; i < children; i++) {
+		child_id = fork();
+
+		if (child_id > 0) {
+			/* PID is being reused. */
+			if (child_id == parent_pid) {
+				fprintf(stderr, "!!! pid %d reused (see ps)\n",
+					child_id);
+				/* Keep this child alive enough time to hit the
+				 * csb error signal. */
+				sleep(300);
+			}
+			continue;
+		} else if (child_id == 0) {
+			pid_t pid = getpid();
+			fprintf(stderr, "%d: child(%d)\n", i, (int) pid);
+			/* Wait enough time to fork all children. */
+			sleep(seconds);
+			fprintf(stderr, "%d: child(%d) exits\n", i, (int) pid);
+			exit(0);
+		} else if (child_id == -1) {
+			fprintf(stderr, "fork failed: %s\n", strerror(errno));
+			exit(-1);
+		}
+	}
+}
+
+int main(int argc, char **argv)
+{
+	int rc, compLen = 1024;
+	char *compr;
+	pthread_t thread;
+	void *ret;
+	long iterations;
+	z_const char hello[] = "hello, hello!";
+
+	if (argc == 2)
+		iterations = atoi(argv[1]);
+	else
+		iterations = 100;
+
+	/* Get parent thread kernel PID. */
+	pid_t my_pid = syscall(SYS_gettid);
+	fprintf(stderr, "parent thread started: pid: %d\n", (int) my_pid);
+
+	memset(&c_stream, 0, sizeof(c_stream));
+	c_stream.zalloc = Z_NULL;
+	c_stream.zfree = Z_NULL;
+	c_stream.opaque = Z_NULL;
+
+	compr = malloc(compLen);
+	assert(compr != NULL);
+
+	rc = pthread_create(&thread, NULL, comp_file_multith,
+			    (void *) &iterations);
+	if (rc != 0) {
+		fprintf(stderr, "error: pthread_create %d\n", rc);
+		return rc;
+	}
+
+	/* Run the compression and leave the window open. */
+	rc = pthread_join(thread, &ret);
+	if (rc != 0) {
+		fprintf(stderr, "error: pthread cannot be joined %p"
+			"\n", ret);
+		return rc;
+	}
+
+	/* Fork many children to see if any use the same PID as the parent. */
+	fork_children((int) my_pid);
+	/* Wait for all children to run. */
+	sleep(60);
```

Comment:
If the goal is to wait for all children processes, you can use `wait()`, calling it in a loop until it returns -1 and `errno == ECHILD`.

Reply:
Does it triggers when the child is sleeping?
Because we need to continue while the child that reused the PID is sleeping, not killed. (line 85)

Reply:
> Does it triggers when the child is sleeping?

@rzinsly No, it does not.

> we need to continue while the child that reused the PID is sleeping, not killed. (line 85)

In that case, I'm afraid there is a bug in this code.
The parent could exit before all its children complete without calling `wait()`, leaving zombie processes behind.

Reply:
It shouldn't 60 seconds proved enough to wait for all children to finish or to get to the point that we hit the bug, in this case it will segfault later.
I added the wait() in loop at the end of the test for safety anyway.